File: 4/day4.go
package day4

import (
	"fmt"
	"math"
	"os"
	"strconv"
	"strings"
)

func check(e error) {
	if e != nil {
		panic(e)
	}
}

func readFile(filename string) string {
	dat, err := os.ReadFile(filename)
	check(err)
	return string(dat)
}

func to_int(char string) int {
	value, err := strconv.Atoi(char)
	check(err)
	return value
}

func findMatching(row string) int {
	_data := strings.Split(row, ":")[1]
	data := strings.Split(_data, "|")
	_winning, _having := data[0], data[1]
	winning := strings.Split(_winning, " ")
	having := strings.Split(_having, " ")

	matching := 0

	for _, _x := range winning {
		_x = strings.TrimSpace(_x)
		if _x == "" {
			continue
		}

		for _, _y := range having {
			_y = strings.TrimSpace(_y)
			if _y == "" {
				continue
			}

			x := to_int(_x)
			y := to_int(_y)

			if x == y {
				matching++
			}

		}
	}

	return matching
}

func processRow(row string) int {
	// 0 matching -> 0 points
	// 1 matching -> 1 points
	// 2 matching -> 2 points
	// 3 matching -> 4 points
	// 4 matching -> 8 points
	// f(n)=2^(n-1)
	// this formula does NOT work for n = 0
	// but we can just manually check that case

	matching := findMatching(row)

	if matching != 0 {
		fmt.Printf("matching: %d\n", matching)
		// f(n)=2^(n-1)
		n := float64(matching - 1)
		points := math.Pow(2, n)
		return int(points)
	} else {
		return 0
	}
}

type memoized struct {
	f     func(int) int
	cache map[int]int
}

func (m *memoized) memoize(n int) int {
	if val, ok := m.cache[n]; ok {
		return val
	}
	result := m.f(n)
	m.cache[n] = result
	return result
}

func memoize(f func(int) int) *memoized {
	return &memoized{f: f, cache: make(map[int]int)}
}

func _processRow2(row string, rows []string, idx int) int {
	// result: 1 + sum(process(winning))
	matching := findMatching(row)

	result := 1

	for i := 1; i <= matching; i++ {
		new_idx := idx + i

		if new_idx >= len(rows) {
			// out of bounds, skip!
			continue
		}

		result += processRow2(rows[new_idx], rows, new_idx)
	}

	return result
}

func processRow2(row string, rows []string, idx int) int {
	process := func(_idx int) int {
		return _processRow2(row, rows, _idx)
	}

	mem := memoize(process)

	return mem.memoize(idx)
}

func run2(filename string) int {
	result := 0

	contents := readFile(filename)
	rows := strings.Split(contents, "\n")

	for idx, row := range rows {
		result += processRow2(row, rows, idx)
	}

	return result
}

File: 4/day4.rs
use std::fs;
use std::collections::HashMap;


fn split<'a>(original_string: &'a str, delimiter: &'a str) -> (&'a str, &'a str) {
    let mut parts = original_string.split(delimiter);
    let first_half = parts.next().unwrap_or("");
    let second_half = parts.next().unwrap_or("");

    return (first_half, second_half);
}

fn to_int(txt: &str) -> u32 {
    match txt.parse::<u32>() {
        Ok(n) => n,
        Err(_) => 0,
    }
}

fn find_matching(line: &str) -> u32 {
    let (_, _data) = split(line, ":");
    let (_winning, _having) = split(_data, "|");

    let winning = _winning.split(" ");
    let having = _having.split(" ");

    let mut matching = 0;

    for w in winning {
        if w == "" {
            continue;
        }

        for h in having.clone() {
            if h == "" {
                continue;
            }

            if to_int(h) == to_int(w) {
                matching += 1
            }
        }
    }

    return matching;
}

fn process_line(line: &str) -> u32 {
    let matching = find_matching(line);

    if matching == 0 {
        return 0;
    }

    return u32::pow(2, matching - 1);
}



fn memoize<F, T, U>(func: F) -> impl FnMut(T) -> U
where
    F: Fn(T) -> U,
    T: std::cmp::Eq + std::hash::Hash + Copy,
    U: Copy,
{
    let mut memo = HashMap::new();
    move |arg| {
        if let Some(&result) = memo.get(&arg) {
            return result;
        }
        let result = func(arg);
        memo.insert(arg, result);
        result
    }
}

fn process_line2(line: &str, lines: Vec<&str>, idx: u32) -> u32 {
    let matches = find_matching(line);

    let mut result = 1;

    for i in 1..=matches {
        let new_idx = idx + i;
        let nextline = lines[new_idx as usize];
        result += process_line2(nextline, lines.clone(), new_idx);
    }


    return result;
}

// fn process_line2(line: &str, lines: Vec<&str>, idx: u32) -> u32 {
//     let mut mem = memoize(|_idx| _process_line2(line, lines.clone(), _idx));
//
//     return mem(idx);
// }

fn run(filename: &str) -> u32 {
    let contents = fs::read_to_string(filename)
        .expect("Should have been able to read the file");

    let mut result = 0;

    for line in contents.split("\n") {
        result += process_line(line);
    }

    return result;
}


fn run2(filename: &str) -> u32 {
    let contents = fs::read_to_string(filename)
        .expect("Should have been able to read the file");

    let mut result = 0;

    let lines: Vec<&str> = contents.split("\n").collect();

    for (idx, line) in lines.iter().enumerate() {
        result += process_line2(line, lines.clone(), idx as u32);
    }

    return result;
}

fn main() {
    // let result = run("dev.txt");
    // println!("result: {r}", r = result);
    // assert!(result == 13);

    let result = run2("dev.txt");
    println!("result: {r}", r = result);
    assert!(result == 30);

    let pt1_result = run2("test.txt");
    println!("part 2: {}", pt1_result);
}
